### ISSUE TITLE
Constraint "CSI volume has exhausted its available writer claims": 1 nodes excluded by filter

### ISSUE THREAD

**Issue by gregory112** [hashicorp/nomad#10927]
### Nomad version
Nomad v1.1.2 (60638a086ef9630e2a9ba1e237e8426192a44244)

### Operating system and Environment details
Ubuntu 20.04 LTS

### Issue
Cannot re-plan jobs due to CSI volumes being claimed.
I have seen many variations about this issue. I don't know how to debug it. I use ceph-csi plugin to deploy system job on my two Nomad nodes. This result in two controllers and two ceph-csi nodes. I then create a few volumes using `nomad volume create` command. I then create a job with three tasks that use three volumes. Sometimes, after a while the job may fail, and I stop it. After that when I try to replan the exact same job I get that error.

What confuses me is the warning. It differs every time I run `job plan`. First I saw

```
- WARNING: Failed to place all allocations.
  Task Group "zookeeper1" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper1-data has exhausted its available writer claims": 2 nodes excluded by filter

  Task Group "zookeeper2" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper2-data has exhausted its available writer claims": 2 nodes excluded by filter
```

Then, runnig `job plan` again a few seconds after, I got

```
- WARNING: Failed to place all allocations.
  Task Group "zookeeper1" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper1-datalog has exhausted its available writer claims": 2 nodes excluded by filter

  Task Group "zookeeper2" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper2-datalog has exhausted its available writer claims": 2 nodes excluded by filter
```

Then again,
```
- WARNING: Failed to place all allocations.
  Task Group "zookeeper1" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper1-data has exhausted its available writer claims": 1 nodes excluded by filter
    * Constraint "CSI volume zookeeper1-datalog has exhausted its available writer claims": 1 nodes excluded by filter

  Task Group "zookeeper2" (failed to place 1 allocation):
    * Constraint "CSI volume zookeeper2-datalog has exhausted its available writer claims": 2 nodes excluded by filter
```

I have three groups: zookeeper1, zookeeper2, and zookeeper3, each using two volumes (data and datalog). I will just assume from this log that all volumes are non-reclaimable.

This is the output of `nomad volume status`.

```
Container Storage Interface
ID                           Name                Plugin ID  Schedulable  Access Mode
zookeeper1-data     zookeeper1-data     ceph-csi   true         single-node-writer
zookeeper1-datalog  zookeeper1-datalog  ceph-csi   true         single-node-writer
zookeeper2-data     zookeeper2-data     ceph-csi   true         single-node-writer
zookeeper2-datalog  zookeeper2-datalog  ceph-csi   true         single-node-writer
zookeeper3-data     zookeeper3-data     ceph-csi   true         <none>
zookeeper3-datalog  zookeeper3-datalog  ceph-csi   true         <none>
```
It says that they are schedulable. This is the output of `nomad volume status zookeeper1-datalog`:

```
ID                   = zookeeper1-datalog
Name                 = zookeeper1-datalog
External ID          = 0001-0024-72f28a72-0434-4045-be3a-b5165287253f-0000000000000003-72ec315b-e9f5-11eb-8af7-0242ac110002
Plugin ID            = ceph-csi
Provider             = cephfs.nomad.example.com
Version              = v3.3.1
Schedulable          = true
Controllers Healthy  = 2
Controllers Expected = 2
Nodes Healthy        = 2
Nodes Expected       = 2
Access Mode          = single-node-writer
Attachment Mode      = file-system
Mount Options        = <none>
Namespace            = default

Allocations
No allocations placed
```

It says there, there are no allocations placed.

### Reproduction steps
This is unfortunately flaky. But most likely happen due to job failing and then stopped and then replanned. This persists even after I purge the job with `nomad job stop -purge`. No, doing `nomad system gc`, `nomad system reconcile summary`, or restarting Nomad does not work.

#### Expected Result
Should be able to reclaim the volume again without having to detach or deregister -force and register again. I created the volumes using `nomad volume create` so those volumes have their external IDs all generated. There are 6 volumes and 2 nodes, I don't want to type detach 12 times everytime this happens (this happens so frequently).

#### Actual Result
See error logs above.

### Job file (if appropriate)
I have three groups (zookeeper1, zookeeper2, zookeeper3) each having volume stanza like this (each with their own volumes, this one is for zookeeper2):
```
    volume "data" {
      type = "csi"
      read_only = false
      source = "zookeeper2-data"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"

      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }
    volume "datalog" {
      type = "csi"
      read_only = false
      source = "zookeeper2-datalog"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"

      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }
```

All groups have `count = 1`.

**Comment by bfqrst** [comment]
[Nomad 1.1.2]

I have similar issues using the AWS EBS plugin. And I wholeheartedly share the sentiment that it's really hard to reproduce or debug, which is why I've been reluctant to opening an issue. In my case Nomad workers are spot instances which come and go. Non CSI jobs get rescheduled fine, but those which have CSI volumes attached tend to go red on the first reschedule and then eventually succeed. It's almost like the CSI mechanism needs more time to do it's thing, before the job gets restarted...

**Comment by gregory112** [comment]
I have been investigating it for a while.
A manual solution that works is to deregister the volume with -force, or at least detaching the volume with `nomad volume detach`. Now `nomad volume detach` is probably the most straightforward approach. I learn that what it actually do is to unmount the mount point of the particular volume in the particular node. And this sometimes hang. When hanging, I see an `umount` process in the node hanging indefinitely. Network connectivity is never reliable, so I think cases like this needs attention.

I don't know what Nomad does with unused volumes after the job is stopped. Does Nomad instruct the plugin to unmount the volumes from nodes? What if it fails to do so then? Maybe a timeout, or more log messages, or anything should be added in `nomad alloc status`.

`umount` hangs often especially in the case of mounting network file systems like this. And when umount hangs, `nomad volume detach` hangs too. The solution in my case is to actually kill the umount process in nodes and retry with `umount -f`, but I don't know if Nomad can do this, as I think this might be managed by the CSI plugin. But due to the variations that exist in a lot different CSI plugins, I think cases like this should be handled by Nomad somehow.

**Comment by konsti** [comment]
We have the very same issue with Nomad 1.1.3 and AWS EBS CSI Plugin v1.2.0.
We need to revert to host_volumes as this issue threatens our cluster health when important stateful services like Prometheus can't restart correctly.

The volume is unmounted currently in AWS and shows `state = "Available"`, so it's clearly an internal Nomad state issue.

**Comment by gregory112** [comment]
Well this is bad, as I don't think the issue is solved by Nomad garbage collector. When a volume hangs in one Nomad node, Nomad will just allocate the job to another node. And when it hangs there, it will allocate the job to another node again. Imagine if we have 100 Nomad nodes and the volumes stuck in some of them.

**Comment by JanMa** [comment]
I am seeing the same issues using the `gcp-compute-persistent-disk-csi-driver` plugin and Nomad `v1.1.1`. Interestingly the [Read Volume](https://www.nomadproject.io/api-docs/volumes#read-volume) API endpoint shows the correct amount of writers but the [List Volumes](https://www.nomadproject.io/api-docs/volumes#list-volumes) endpoint shows too many `CurrentWriters`. I guess this is what's stopping the reassignment of the volume

**Comment by Thunderbottom** [comment]
We've faced this issue as well. Seems like Nomad fails to realise that there are not allocations present for the volume.

From what we could make out, both AWS and the CSI plugin report that the volume is available to mount:

```
$ nomad volume status -verbose -namespace=workspace workspace
ID                   = workspace
Name                 = workspace
External ID          = vol-xxxxxxxxxxxxx
Plugin ID            = aws-ebs0
Provider             = ebs.csi.aws.com
Version              = v0.10.1
Schedulable          = true
Controllers Healthy  = 1
Controllers Expected = 1
Nodes Healthy        = 2
Nodes Expected       = 2
Access Mode          = single-node-writer
Attachment Mode      = file-system
Mount Options        = <none>
Namespace            = workspace
```

And the AWS console reports the volume is "Available".

Whereas in the Nomad UI, the "Storage" tab reports multiple `#Allocs` for the aforementioned volume, and clicking on the volume shows that there are neither read nor write mounts, which potentially points to my hunch that Nomad somehow isn't aware that the volume has been detached by the driver and is not in use by any of the allocations. Here's a gist of logs from the CSI controller for the same:

```
I0816 06:26:49.482851       1 controller.go:329] ControllerUnpublishVolume: called with args {VolumeId:vol-XXXXXXXXXX NodeId:i-YYYYYYYYYYY Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0816 06:26:56.194278       1 controller.go:346] ControllerUnpublishVolume: volume vol-XXXXXXXXXX detached from node i-YYYYYYYYYYY
```

One thing I have noticed is that on a newly registered CSI volume, the metadata shows up as:

```
Access Mode          = <none>
Attachment Mode      = <none>
```

And while on the volume that is "Schedulable" (not really, in terms of Nomad) it shows up as:

```
Access Mode          = single-node-writer
Attachment Mode      = file-system
```

Could be pointing to something, perhaps?

The only solution that has worked for us so far is to `-force` deregister the volume and re-register it. Would love to have this issue solved. Let me know if I can help with any more information.

**Comment by lgfa29** [comment]
Thank you all for the detailed information (keep them coming if you have more!).

It seems like this issue happens sporadically when allocations that use a volume are restarted. I will try to create a high churn environment and see if I can reproduce it.

**Comment by lgfa29** [comment]
Hi everyone, just a quick update. I've been running a [periodic job](https://gist.github.com/lgfa29/b1ffa1b99974d6e82f94a40a5a28f843) for a couple of days now, and so far I haven't seen any issues.

In hindsight I should've probably used a `service` job since they are likely more common they have some different scheduling logic when compared to `batch`.

For those who have seen this issue, do you have any kind of `update` block set? This could affect how new allocations are created and I want to make sure I have a good reproduction environment.

Thanks!

**Comment by JanMa** [comment]
Hey @lgfa29 you can try out the following job. It's a basic `prometheus` deployment without any special `update` block.
I regularly have issues when trying to update this for, for example to add more resources.
```hcl
job "prometheus" {
  datacenters = ["dc1"]
  type = "service"

  group "monitoring" {
    count = 2

    constraint {
      operator = "distinct_hosts"
      value = "true"
    }

    volume "data" {
      type            = "csi"
      source          = "prometheus-disk"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"
      per_alloc       = true
    }

    network {
      port "http" {
        static = 9090
      }
    }

    service {
      name = "prometheus2"
      tags = ["prometheus2"]
      task = "prometheus"
      port = "http"

      check {
        type = "http"
        port = "http"
        path = "/-/ready"
        interval = "10s"
        timeout = "5s"
      }
    }

    task "prometheus" {
      driver = "docker"
      user = "root"

      volume_mount {
        volume      = "data"
        destination = "/prometheus"
      }

      resources {
        memory = 1024
        cpu = 1024
      }

      template {
        data = <<EOT
---
global:
  scrape_interval: 10s
  external_labels:
    __replica__: "{{ env "NOMAD_ALLOC_ID" }}"
scrape_configs:
  - job_name: "prometheus"
    scrape_interval: 10s
    consul_sd_configs:
      - server: "{{ env "attr.unique.network.ip-address" }}:8500"
        services:
          - prometheus2
    relabel_configs:
      - source_labels: ["__meta_consul_node"]
        regex: "(.*)"
        target_label: "node"
        replacement: "$1"
      - source_labels: ["__meta_consul_service_id"]
        regex: "(.*)"
        target_label: "instance"
        replacement: "$1"
      - source_labels: ["__meta_consul_dc"]
        regex: "(.*)"
        target_label: "datacenter"
        replacement: "$1"
  # Nomad metrics
  - job_name: "nomad_metrics"
    consul_sd_configs:
      - server: "{{ env "attr.unique.network.ip-address" }}:8500"
        services: ["nomad-client", "nomad"]
    relabel_configs:
      - source_labels: ["__meta_consul_tags"]
        regex: "(.*)http(.*)"
        action: keep
      - source_labels: ["__meta_consul_service"]
        regex: "(.*)"
        target_label: "job"
        replacement: "$1"
      - source_labels: ["__meta_consul_node"]
        regex: "(.*)"
        target_label: "node"
        replacement: "$1"
      - source_labels: ["__meta_consul_service_id"]
        regex: "(.*)"
        target_label: "instance"
        replacement: "$1"
      - source_labels: ["__meta_consul_dc"]
        regex: "(.*)"
        target_label: "datacenter"
        replacement: "$1"
    scrape_interval: 5s
    metrics_path: /v1/metrics
    params:
      format: ["prometheus"]
  # Consul metrics
  - job_name: "consul_metrics"
    consul_sd_configs:
      - server: "{{ env "attr.unique.network.ip-address" }}:8500"
        services: ["consul-agent"]
    relabel_configs:
      - source_labels: ["__meta_consul_tags"]
        regex: "(.*)http(.*)"
        action: keep
      - source_labels: ["__meta_consul_service"]
        regex: "(.*)"
        target_label: "job"
        replacement: "$1"
      - source_labels: ["__meta_consul_node"]
        regex: "(.*)"
        target_label: "node"
        replacement: "$1"
      - source_labels: ["__meta_consul_service_id"]
        regex: "(.*)"
        target_label: "instance"
        replacement: "$1"
      - source_labels: ["__meta_consul_dc"]
        regex: "(.*)"
        target_label: "datacenter"
        replacement: "$1"
    scrape_interval: 5s
    metrics_path: /v1/agent/metrics
    params:
      format: ["prometheus"]
EOT
        destination = "local/prometheus.yml"
      }
      config {
        image = "quay.io/prometheus/prometheus"
        ports = ["http"]
        args = [
          "--config.file=${NOMAD_TASK_DIR}/prometheus.yml",
          "--log.level=info",
          "--storage.tsdb.retention.time=1d",
          "--storage.tsdb.path=/prometheus",
          "--web.console.libraries=/usr/share/prometheus/console_libraries",
          "--web.console.templates=/usr/share/prometheus/consoles"
        ]
      }
    }
  }
}
```

**Comment by gregory112** [comment]
> For those who have seen this issue, do you have any kind of `update` block set? This could affect how new allocations are created and I want to make sure I have a good reproduction environment.

Mine has.
```
  update {
    health_check = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "15m"
    progress_deadline = "20m"
  }
```

Also you can try to fail the job. For example due to pull fail or running fail to the point the deployment exceeds deadline. This is what really makes the error frequently show up.

Also I have two master nodes. Don't know if this contributes to this but does anyone use two master nodes too for this? Or maybe more? Maybe a race condition error between master nodes? Split brain perhaps?

**Comment by lgfa29** [comment]
Thanks for the sample jobs @JanMa and @gregory112. 

It seems like `update` is not part of the problem, so I create this job that has a 10% chance of failing. I will leave it running and see if the issue is triggered.

```hcl
job "random-fail" {
  datacenters = ["dc1"]
  type        = "service"

  group "random-fail" {
    volume "ebs-vol" {
      type            = "csi"
      read_only       = false
      source          = "ebs-vol"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"

      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }

    task "random-fail" {
      driver = "docker"

      config {
        image   = "alpine:3.14"
        command = "/bin/ash"
        args    = ["/local/script.sh"]
      }

      template {
        data = <<EOF
#!/usr/bin/env bash

while true;
do
  echo "Rolling the dice..."

  n=$(($RANDOM % 10))
  echo "Got ${n}!"

  if  [[ 0 -eq ${n} ]];
  then
    echo "Bye :wave:"
    exit 1;
  fi

  echo "'Til the next round."
  sleep 10;
done
EOF

        destination = "local/script.sh"
      }

      volume_mount {
        volume      = "ebs-vol"
        destination = "/volume"
        read_only   = false
      }
    }
  }
}
```

**Comment by raven-oscar** [comment]
For some reason nomad thinks that volume is still in use while it is not. nomad volume deregister returns "Error deregistering volume: Unexpected response code: 500 (rpc error: volume in use: nessus)". nomad volume deregister -force followed by nomad system gc and then registering it again seems to help.

**Comment by lgfa29** [comment]
I think I was able to reproduce this with a self-updating job:

```hcl
job "update" {
  datacenters = ["dc1"]
  type        = "service"

  group "update" {
    volume "ebs-vol" {
      type            = "csi"
      read_only       = false
      source          = "ebs-vol2"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"

      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }

    task "update" {
      driver = "docker"

      config {
        image   = "alpine:3.14"
        command = "/bin/sh"
        args    = ["/local/script.sh"]
      }

      template {
        data = <<EOF
#!/usr/bin/env bash

apk add curl jq

while true;
do
  sleep 10
  jobspec=$(curl http://172.17.0.1:4646/v1/job/${NOMAD_JOB_ID})
  cpu=$(echo $jobspec | jq '.TaskGroups[0].Tasks[0].Resources.CPU')
  if [ $cpu -eq 500 ]
  then
    cpu=600
  else
    cpu=500
  fi
  new_jobspec=$(echo $jobspec | jq ".TaskGroups[0].Tasks[0].Resources.CPU = ${cpu}")
  echo $new_jobspec | jq '{"Job":.}' | curl -H "Content-Type: application/json" -X POST --data @- http:/172.17.0.1:4646/v1/job/${NOMAD_JOB_ID}
done
EOF

        destination = "local/script.sh"
      }

      volume_mount {
        volume      = "ebs-vol"
        destination = "/volume"
        read_only   = false
      }

      resources {
        cpu    = 500
        memory = 256
      }
    }
  }
}
```

This job will usually fail first, and then work.
<img width="1189" alt="image" src="https://user-images.githubusercontent.com/775380/130647607-7d15d9fb-410d-4f27-bf51-6bb20d6960f1.png">

It's not exactly the same error message, but maybe it's related:

> failed to setup alloc: pre-run hook "csi_hook" failed: claim volumes: rpc error: controller publish: attach volume: controller attach volume: CSI.ControllerAttachVolume: controller plugin returned an internal error, check the plugin allocation logs for more information: rpc error: code = Internal desc = Could not attach volume "vol-0fa6e88e4c3b55d29" to node "i-02396ab34ed318944": attachment of disk "vol-0fa6e88e4c3b55d29" failed, expected device to be attached but was detaching

Looking at the volume details page, I see the previous allocation still listed under `Write Allocations` for a while, even after it's marked as `Complete`. I don't know if this is related, since another alloc is able to start, but may be be a pointer in the right direction
<img width="1190" alt="image" src="https://user-images.githubusercontent.com/775380/130648198-f7a49d7b-067e-4d47-872a-8aea4a799d1a.png">

**Comment by latifrons** [comment]
Same issue on my cluster with EFS plugin (docker: amazon/aws-efs-csi-driver:v1.3.3)
It usually happens after a restart of some node. No Write Allocations, No Read Allocations, however failed to mount because CSI volume has exhausted its available writer claims.

One thing needs to be mentioned is that, right after force deregistering the volume, my Node Health recovered from (4/5) to (5/5). EFS plugin cannot be deployed to that node before deregistering.

**Comment by aroundthfur** [comment]
We have the exact same issue with the Linstor CSI driver (https://linbit.com/linstor/).
Our use-case is to be able to fail-over a job when one server or a DC dies. We are still in the implementation phase so we were intentionally killing servers where nomad jobs are running to reproduce the fail-over use-cases.

Now, it works a few times and nomad does a proper job, but after a random number of re-schedules it fails with the same error message and the same behavior where in the Nomad UI volume list the volume has an allocation set, but when we go into the detailed view there is no allocs present. Basically the same behavior that @Thunderbottom reported.

**Comment by MagicRB** [comment]
Yep, I'm running a really basic setup with just one server/client and one client and this happens to me too, I need to `-force` deregister the volume and then re-register it to fix it

**Comment by latifrons** [comment]
Just a kind reminder for those who are using AWS EFS:
Make sure you are using the suggested mount options given by AWS.
https://docs.aws.amazon.com/efs/latest/ug/efs-mount-helper.html

especially noresvport flag.

Or it will cause infinite time wait io (100% wa for top) once tcp connection breaks.

We didn't observe any more problem after correcting the mount options.

> Same issue on my cluster with EFS plugin (docker: amazon/aws-efs-csi-driver:v1.3.3)
> It usually happens after a restart of some node. No Write Allocations, No Read Allocations, however failed to mount because CSI volume has exhausted its available writer claims.
> 
> One thing needs to be mentioned is that, right after force deregistering the volume, my Node Health recovered from (4/5) to (5/5). EFS plugin cannot be deployed to that node before deregistering.

**Comment by tomiles** [comment]
Same issue on our v1.1.4 cluster using Ceph RBD and cepfs CSI plugins

**Comment by gregory112** [comment]
This may not be a network issues, as even though I deployed Ceph and Nomad on the same node, mounting to localhost, the issue still persist. nomad detach works fine without any problem as network issues are now out of the way. I also used a single Nomad server, but this issue still exist. The weird thing about it is when I have three instances (zoo1, zoo2, zo3) all asking for their own volumes, only some volumes (not all) have issues. This could be a concurrency problem? Race condition perhaps?

**Comment by crestonbunch** [comment]
I've noticed my CSI plugins (both the node and controllers) are frequently restarting maybe 3-4 times a day. I don't know why yet. I suspect this sometimes leaves the volumes in a bad state and causes this error.

**Comment by keslerm** [comment]
I'm hitting the same thing using seaweedfs. Ultimately i think a network issue was causing the volumes to disconnect but nomad was never releasing them internally until i deregistered/reregistered them

**Comment by gregory112** [comment]
Any update on this? @lgfa29

**Comment by vgeddes** [comment]
Same issue on our v1.1.5 cluster using AWS EBS volumes. We have to manually force deregister and register the affected volumes.   Happens very frequently.

**Comment by lgfa29** [comment]
> Any update on this? @lgfa29

Hi @gregory112, no updates yet. I will let you know once we do have any progress to share here.

**Comment by JanMa** [comment]
I have been digging through the Nomad source code and found a possible
work-around that works well for my use-case. If the volume gets registered as a
`multi-node-multi-writer` volume instead of a `single-node-writer`
volume, the parts of the source code that I think are buggy are just
skipped. And because I set the `per_alloc` flag to `true` I am rather
confident that Nomad or the CSI driver will not actually try to attach a
volume that is in use to another alloc/task.

This might have some other unintended consequences but for now it allows
me to deploy updated versions of the job again without having to
re-register the volume every time.

**Comment by m1keil** [comment]
Another +1 here. We are having these issues in both dev and prod with AWS EBS.

Nomad v1.1.1 with EBS CSI driver v1.1.0.
No special `update {}` stanza (using default).

As others described, it looks like Nomad "forgets" the volume is no longer allocated. In my case, I even had a job that was not running for weeks, and when I tried to redeploy, Nomad still complained about "CSI volume <name> has exhausted its available writer claims".

Inspecting volume status showed the volume as "No allocations placed" and check AWS confirmed the volume isn't mounted to any instance.

So far the only way to overcome this is to deregister and re-register to volume.

**Comment by crestonbunch** [comment]
I'm seeing a lot of errors in logs of the form "error releasing volume claims [...] Permission denied".

```
ubuntu@ip-10-0-2-23:~$ journalctl -u nomad | grep "error releasing volume claims" -A 2 | tail
Nov 09 00:28:10 ip-10-0-2-23 nomad[24326]: "
Nov 09 00:36:15 ip-10-0-2-23 nomad[24326]:     2021-11-09T00:36:15.221Z [ERROR] nomad.volumes_watcher: error releasing volume claims: namespace=default volume_id=auth-scylladb[1] error="1 error occurred:
Nov 09 00:36:15 ip-10-0-2-23 nomad[24326]:         * Permission denied
Nov 09 00:36:15 ip-10-0-2-23 nomad[24326]: "
Nov 09 00:48:20 ip-10-0-2-23 nomad[24326]:     2021-11-09T00:48:20.202Z [ERROR] nomad.volumes_watcher: error releasing volume claims: namespace=default volume_id=auth-scylladb[1] error="1 error occurred:
Nov 09 00:48:20 ip-10-0-2-23 nomad[24326]:         * Permission denied
Nov 09 00:48:20 ip-10-0-2-23 nomad[24326]: "
Nov 09 01:08:27 ip-10-0-2-23 nomad[24326]:     2021-11-09T01:08:27.605Z [ERROR] nomad.volumes_watcher: error releasing volume claims: namespace=default volume_id=auth-scylladb[1] error="1 error occurred:
Nov 09 01:08:27 ip-10-0-2-23 nomad[24326]:         * Permission denied
Nov 09 01:08:27 ip-10-0-2-23 nomad[24326]: "
```

**Comment by rwojsznis** [comment]
Hi all 👋 Was this issue ever fixed? I'm running nomad 1.1.6 with [hetzner csi driver](https://github.com/hetznercloud/csi-driver) 1.6.0 and when it works it's great, but when it breaks it's a pure chaos - stuck controller allocation (even tho csi plugin was successfully started by docker and throws no errors), nomad throws 'exhausted claim' error once csi driver allocation succeeds - today I had to force deregister & register all volumes to get it working again.

Any way I can help debugging this?

**Comment by samed** [comment]
Another +1 from here. I've tried both EBS and ceph and hit this issue.  Is this related to https://github.com/hashicorp/nomad/issues/10833 @tgross @lgfa29

**Comment by MagicRB** [comment]
I don't want to be mean towards the Nomad team, but this is a really big issue in my opinion, it basically disqualifies CSI for any serious HA deployments (not that my person website is HA), I'm purposefully avoiding CSI on a few key services to make sure they don't randomly die. Just my two cents, sucks that this is an issue... Nomad is otherwise a stellar system.

**Comment by crestonbunch** [comment]
+1. Nomad is truly an amazing alternative to Kubernetes. But at times I have considered switching just because of this issue. In the meantime I have stopped using CSI altogether.

My use case is simply to mount an AWS EBS volume for some databases. I can do that without CSI by running a script as a pre- and post-start lifecycle tasks. If anyone is interested, [here](https://gist.github.com/crestonbunch/5f5a2aed4f103da3a280147f6e7d730c) are the scripts I use. 

With CSI, tasks died every day and failed to restart. With this solution not a single task has failed.

**Comment by raven-oscar** [comment]
+1 CSI at the moment can be viewed extremely unreliable because of this bug.

**Comment by schmichael** [comment]
The Nomad team at HashiCorp hears you! Sorry we haven't shipped a fix yet, but it's on our roadmap. In the mean time I've pinned the issue to raise its visibility.

In the mean time can anyone confirm how viable the two workarounds mentioned are? I've seen:

1. Use multi-node-multi-writer access mode - May not work with all drivers or jobs - https://github.com/hashicorp/nomad/issues/10927#issuecomment-935747605
2. Force detach the volume: https://github.com/hashicorp/nomad/issues/10927#issuecomment-917676568

**Comment by trilom** [comment]
I will say from my experience dealing with wanting to use ceph with nomad has been kinda bad.

One thing that has caused a lot of pain is following guidance in this area https://github.com/hashicorp/nomad/tree/main/demo/csi/ceph-csi-plugin, for example using the `--instance-id` flag to the instance per alloc_id will cause the controller to create an omap key per alloc, such that whenever an alloc of controller dies it creates a new omap ID and cannot find previous volumes.

When coupled with the pain of "deregistering forcefully" volumes and re-registering them as a solution and the controller has changed causes it to not know in ceph where volumes are, alloc XYZ:VOLUME1 is not the same as alloc ABC:VOLUME1.

**Comment by keslerm** [comment]
@schmichael force de-registering the volume and re-creating it is how I've been dealing with it using the Linode CSI plugin

**Comment by m1keil** [comment]
@schmichael multi-node-multi-writer access mode isn't supported by EBS so cannot be applied.

**Comment by JohnKiller** [comment]
> using the `--instance-id` flag to the instance per alloc_id will cause the controller to create an omap key per alloc, such that whenever an alloc of controller dies it creates a new omap ID and cannot find previous volumes.

@trilom are you saying that the documentation is wrong and that flag shouldn't be used or is this bug that is preventing that from working the way it's intended to? I'm also running ceph and the information on the internet is very scarce.

**Comment by gregory112** [comment]
> I will say from my experience dealing with wanting to use ceph with nomad has been kinda bad.
> 
> One thing that has caused a lot of pain is following guidance in this area https://github.com/hashicorp/nomad/tree/main/demo/csi/ceph-csi-plugin, for example using the `--instance-id` flag to the instance per alloc_id will cause the controller to create an omap key per alloc, such that whenever an alloc of controller dies it creates a new omap ID and cannot find previous volumes.
> 
> When coupled with the pain of "deregistering forcefully" volumes and re-registering them as a solution and the controller has changed causes it to not know in ceph where volumes are, alloc XYZ:VOLUME1 is not the same as alloc ABC:VOLUME1.

I still use `${node.unique.name}-nodes` instance-id for Ceph plugin node and `-controller` for controller and use job `type = "system"`.

Yet, I agree with you, using Ceph coupled with CSI and Nomad gave me a bad experience. On the brightside, the `multi-node-multi-writer` workaround works for me currently, but still the system needs manual intervention as the whole system does not survive reboot. My whole system (Nomad + Consul + Ceph) are in the same host, for testing, and when the system is rebooted, Ceph and Nomad run normally but the plugin does not. The controller and node run but the plugin is not registered in Nomad, or if it's registered, it shows no node in `nomad plugin status`, although all nodes are shown to be running in `nomad job status`. Probably it hangs because the plugin starts first before Ceph, but even if it does, it stays like that forever, so I have to manually restart the job again. I am still investigating this and so have not opened a new bug yet. I wonder if anyone see the same issue as well.

So yes, I agree with you. It is quite unreliable.

**Comment by trilom** [comment]
> > using the `--instance-id` flag to the instance per alloc_id will cause the controller to create an omap key per alloc, such that whenever an alloc of controller dies it creates a new omap ID and cannot find previous volumes.
> 
> @trilom are you saying that the documentation is wrong and that flag shouldn't be used or is this bug that is preventing that from working the way it's intended to? I'm also running ceph and the information on the internet is very scarce.

Check the pool you are creating volumes against with this command `rados ls -p RBD_POOL  | grep csi.volume` this will give you all the omaps created with the defaults.  You should see one named `csi.volumes.default` if you don't assign `instance-id` but if you assign it then this creates an omap key for `csi.volumes.INSTANCE_ID`.  If you ask the omap key what it's value is with `rados listomapvals -p RBD_POOL csi.volumes.default` then you can see that it relates the volume labels's(in nomad) to volume labels's(in ceph).

Consider a scenario where the controller node changes, then this omap key is recreated and doesn't know the ID's of the volumes.  My best idea to use this is if you have a pool with multiple nomad clusters claiming volumes then they both won't use `csi.volumes.default` where volume ID's will reside in the same omap K/V you can have them as `csi.volumes.CLUSTER1` and `csi.volumes.CLUSTER2` by assigning `--instance-id=CLUSTER1`.  This is my experience with `canary` tag of ceph-csi.

**Comment by trilom** [comment]
I think my main issue that occurs now for me is what happens if a volume is unable to connect, it doesn't handle this well.  Eventually a container will linger that cannot be killed (docker cannot kill it).  When you check the host, there are kernel errors complaining about the connection to mon and sometimes osd.

I have not found a way to kill that process without killing the node.  Investigating the node you can see that the rbd mount still exists so it must not have been unmounted and unmapped successfully.  After killing the node, and deregistering and recreating the volume another node (or the same after rebooting) can claim the volume.

There is no graceful way to handle this other than working against a known good pattern of volume claims, introducing a ton of process, or switch to another workload orchestrator to have easy volumes.

**Comment by tgross** [comment]
Hi folks, can we split off the Ceph demo discussion to a new GitHub issue? I suspect the problem is just that the demo is only a demo, and the job spec assumed a single node ([ref](https://github.com/hashicorp/nomad/tree/main/demo/csi/ceph-csi-plugin#running-ceph-in-vagrant)):
> For demonstration purposes only, you can run Ceph as a single container Nomad job on the Vagrant VM managed by the Vagrantfile at the top-level of this repo.

But let's open a new issue for that so that it gets the attention it deserves!

For folks who are interested in this issue, which is about the unfortunate writer claims bug, as @schmichael said it's on our roadmap and I'm back on the case to fix this and some of the other outstanding issues around CSI to finally push it out of beta. We can't give a timeline but I'll update here and on other open issues regularly as we squash the remaining bugs. Thanks for your patience, all!

**Comment by dcarbone** [comment]
@tgross Point two as listed by @Ramblurr is _extremely_ common.  As an example, here is the cluster I'm using to test out csi for stuff in my home:

![2021-12-12 15 27 30](https://user-images.githubusercontent.com/1392439/145730251-b3ab595d-3330-4ae5-ac65-8626d482ec12.gif)

This is after a single cluster-wide reboot as I replaced / upgrading some switching gear.  To shut down, I did the following:

1. Drain all non-system jobs from each client, one at a time
2. Drain all system jobs from each client, one at a time
3. Stop the nomad client agent before shutting down the host, one at a time.
4. Stop the nomad server agent before shutting down the host, one at a time.

Even with this care, two volumes are now useless to me (`syncthing-config` and `syncthing-data`) in a literal sense.  I cannot deregister them as one thing thinks their allocated, and i cannot detach them as another thing thinks there are no allocations.

I believe this problem is caused by a bunch of places only checking the _length_ of the claims map(s), but not the actual _contents_ of said map(s).

Some examples of this:
1. https://github.com/hashicorp/nomad/blob/main/nomad/structs/csi.go#L359
2. https://github.com/hashicorp/nomad/blob/main/nomad/structs/csi.go#L414

This is a problem as there are several places in the codebase where the map is _written_ to with a `nil` value:

1. https://github.com/hashicorp/nomad/blob/main/nomad/state/state_store.go#L2358
2. https://github.com/hashicorp/nomad/blob/main/nomad/structs/csi.go#L574

I haven't had a chance to dig through the logic too much, but it looks like it's expected for the gc routine to clean these up, but it seems like its pretty easy for things to end up in a permanently broken state.

As you can see in the gif, I went ahead and created _another_ csi volume to replace the syncthing_config volume (and switched to iscsi, because why not) and I'm going to have to do the same thing for the data volume to get things back into a functional state.

**Comment by tgross** [comment]
Hi @dcarbone yup, that's more-or-less the area I'm digging into. See also https://github.com/hashicorp/nomad/issues/8734 and https://github.com/hashicorp/nomad/issues/10052

**Comment by dcarbone** [comment]
Just as a bump, attempting to use terraform to update-in-place my syncthing job has resulted in my _new_ `syncthing-config-iscsi` volume to become stuck forever :(

**Comment by tgross** [comment]
Just a heads up for folks eagerly watching this one, I've broken this work down into a handful of fixes to investigate and implement and the first is https://github.com/hashicorp/nomad/pull/11776

**Comment by tgross** [comment]
Hi folks, just wanted to give an update here. I've got an environment set up where I can reproduce this bug reasonably reliably using the [`democratic-csi`](https://github.com/democratic-csi/democratic-csi/blob/master/docs/nomad.md) plugin (thanks to the folks at @democratic-csi for having a nice Nomad setup there) with NFS. I've found two clues to where the bug may be:

* When [`(*csiHook) Prerun`](https://github.com/hashicorp/nomad/blob/v1.2.3/client/allocrunner/csi_hook.go#L32) runs, it "claims" the volume at the server. This claim kicks off the controller RPCs, and when it returns we mount the volume via the node RPCs. If the node RPCs fail, we return an error and then the [`Postrun`](https://github.com/hashicorp/nomad/blob/v1.2.3/client/allocrunner/csi_hook.go#L81) hook is supposed to unpublish (disclaim) the volume. But the raft logs I have here show _we never disclaim the volume!_ 
  * That leaves a bunch of write claims sitting around for allocations that are dead, but the claims aren't marked as "past claims" that'll get cleaned up. So that looks like a promising avenue of investigation.
  * I'm also questioning whether it makes sense to claim the volume from the client, instead of marking the claim during plan apply and having the client just drive the controller attachment RPCs. That's a sidebar issue.

* Even in this scenario, I would absolutely expect the volume GC to pick up that the allocations associated with the volume are dead and mark the volume for reaping. But I don't see the [unpublish RPC](https://github.com/hashicorp/nomad/blob/v1.2.3/nomad/volumewatcher/volume_watcher.go#L244) calls from the volume watcher either, which suggests either the GC job is not marking the volume for claim reaping correctly, or the volume watcher's logic is wrong. I have actual raft states here that failed this condition, so as long as it's not weirdly racy it's probably possible for me to feed that through some tests to figure out what's going wrong.

**Comment by dcarbone** [comment]
@tgross perhaps in the short term we could have a way to "force" detach volumes?  I'm racking up quite a number of them :(

FWIW, a _very_ reliable way of triggering this problem is to try to update a job in-place.  Update a `template` stanza, or one of the `service.tags` values.  When I try to do this with Terraform, there is around a 1/3 chance that it'll cause at least one of the volumes to become "stuck".

**Comment by trilom** [comment]
> ...perhaps in the short term we could have a way to "force" detach volumes? I'm racking up quite a number of them :(

@dcarbone - check out the `nomad deregister -force` and `nomad volume create` flow, it's what solves my problems.

**Comment by dcarbone** [comment]
@trilom thanks for the tip!

**Comment by tgross** [comment]
> FWIW, a _very_ reliable way of triggering this problem is to try to update a job in-place. Update a `template` stanza, or one of the `service.tags` values. When I try to do this with Terraform, there is around a 1/3 chance that it'll cause at least one of the volumes to become "stuck".

Thanks @dcarbone. This is particularly interesting because the client-side hook doesn't run _at all_ when we do an in-place update of the allocation (which it's a different kind of bug on its own, and why I opened https://github.com/hashicorp/nomad/issues/11786). If the allocation isn't being rescheduled I wouldn't expect the CSI claims to change, so that helps narrow down where the behavior is going wrong. Or adds a new place where it's going wrong.

**Comment by tgross** [comment]
Ok, I've been away for a couple days but I wanted to give a status update on today's investigation.

* I discovered today that our Controller RPC for `CreateVolume` doesn't seem to be working correctly with `democratic-csi`, or at least not with NFS. The directory on the NFS server isn't being created as I'd expect so the mount fails. This was the cause of the alloc failure in https://github.com/hashicorp/nomad/issues/10927#issuecomment-1013467062. I'll need to debug that as well and see whether it's at our side or the plugin's side (probably ours, tbh). But that's not strictly related to the issue here; it just happened to be what caused the initial alloc failure.

* I tried to reproduce in-place updates causing the issue as @dcarbone reported, but I haven't been able to reproduce that for anything that's truly an in-place update (ex. service name update) and doesn't cause a reschedule.

* There's a method in the state store code [`CSIVolumeDenormalizeTxn`](https://github.com/hashicorp/nomad/blob/v1.2.4/nomad/state/state_store.go#L2494) where we "denormalize" allocations for a volume by querying their current state. We backfill claims there for backwards compatibility with pre 1.1.0 state, but I noticed we're not taking advantage of that moment to ensure that any nil allocations are in the `PastClaims` list as well.

* I was able to get an interesting reproduction where a `single-node-writer` job was stopped + purged, and then a new job claimed the same volume in `multi-node-multi-writer` mode (with 2 allocs). This resulted in the following:

```
$ nomad volume status csi-volume-nfs0
ID                   = csi-volume-nfs0
Name                 = csi-volume-nfs0
External ID          = csi-volume-nfs0
Plugin ID            = org.democratic-csi.nfs
Provider             = org.democratic-csi.nfs
Version              = 1.4.3
Schedulable          = true
Controllers Healthy  = 2
Controllers Expected = 2
Nodes Healthy        = 3
Nodes Expected       = 3
Access Mode          = single-node-writer
Attachment Mode      = file-system
Mount Options        = <none>
Namespace            = default

Allocations
ID                                    Node ID                               Task Group  Version  Desired  Status    Created    Modified
399b8755-a1c5-2c2b-8d69-3e635947f154  ef15552c-6a06-bf35-4e6b-4cb37b624e08  web         7        stop     complete  8m47s ago  6m23s ago
93a9b426-ce6b-224c-297d-f9f0e9dc02bf  1d682e06-3776-55cb-38b6-8d53a83cb336  web         0        run      running   2m48s ago  2m37s ago
4e4fdd6e-1202-31f9-bff4-6e377672f106  b05f0de6-7cbd-dc6f-3335-b83ad15b1696  web         0        run      running   2m48s ago  2m37s ago
```

Note that the access and attachment mode are stuck on the old mode, and we see the old allocation floating around. Which means the stopped allocation was being held as a current write claim.

I then drained one of the nodes running an allocation, intentionally neglecting to use `-ignore-system`. As we'd expect (ref https://github.com/hashicorp/nomad/issues/11614), the unmounting step failed because there was no node plugin available.

The job status was very interesting after that:

```
$ nomad job status httpd
...

Placement Failure
Task Group "web":
  * Class "vagrant": 2 nodes excluded by filter
  * Constraint "CSI volume csi-volume-nfs0 has exhausted its available writer claims": 2 nodes excluded by filter

Latest Deployment
ID          = f698e0f7
Status      = successful
Description = Deployment completed successfully

Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
web         2        2       2        0          2022-01-19T16:49:45Z

Allocations
ID        Node ID   Task Group  Version  Desired  Status    Created    Modified
4e4fdd6e  b05f0de6  web         0        stop     complete  2h28m ago  19s ago
93a9b426  1d682e06  web         0        run      running   2h28m ago  2h28m ago
```

The scheduler should only be worried about `multi-node-multi-writer` claims, for which we have no limits. That led me to this block of code in the scheduler at [`feasible.go#L326`](https://github.com/hashicorp/nomad/blob/v1.2.4/scheduler/feasible.go#L326):


```go
if !vol.WriteFreeClaims() {
    // Check the blocking allocations to see if they belong to this job
    for id := range vol.WriteAllocs {
        a, err := c.ctx.State().AllocByID(ws, id)
        if err != nil || a == nil ||
            a.Namespace != c.namespace || a.JobID != c.jobID {
            return false, fmt.Sprintf(
                FilterConstraintCSIVolumeInUseTemplate, vol.ID)
        }
    }
}
```

The `WriteFreeClaims` method assumes that we only ever have one mode set, which isn't the case with our stuck claim. But then the check to make sure that this job is not blocking it's own claims is hitting that `a == nil` condition and failing.

The raft state for these claims looks like this:

```json
{
  "AllocationID": "399b8755-a1c5-2c2b-8d69-3e635947f154",
  "NodeID": "ef15552c-6a06-bf35-4e6b-4cb37b624e08",
  "ExternalNodeID": "",
  "Mode": 1,
  "AccessMode": "single-node-writer",
  "AttachmentMode": "file-system",
  "State": 0
}
{
  "AllocationID": "4e4fdd6e-1202-31f9-bff4-6e377672f106",
  "NodeID": "b05f0de6-7cbd-dc6f-3335-b83ad15b1696",
  "ExternalNodeID": "",
  "Mode": 1,
  "AccessMode": "multi-node-multi-writer",
  "AttachmentMode": "file-system",
  "State": 0
}
{
  "AllocationID": "93a9b426-ce6b-224c-297d-f9f0e9dc02bf",
  "NodeID": "1d682e06-3776-55cb-38b6-8d53a83cb336",
  "ExternalNodeID": "",
  "Mode": 1,
  "AccessMode": "multi-node-multi-writer",
  "AttachmentMode": "file-system",
  "State": 0
}
```


My next steps are:
* Write up a test that passes this volume state through both the GC and the volume reaper to verify that it's not cleaning up the stale claim as we'd expect.
* Test a patch to `CSIVolumeDenormalizeTxn` to ensure nil allocations are in the `PastClaims` list.
* Test a patch to the feasibility checker to ensure nil allocations that don't have corresponding live `PastClaims` aren't blocking write claims.
* I've also got a patch going to do client-driven node unpublish and another moving claims from the client side to the plan applier, but that's probably not going to be ready till next week.
* I'll keep updating this issue frequently so folks know where we're at with this. Thanks again for your patience!

**Comment by Cellivar** [comment]
Very appreciative of the thorough writeup keeping us in the loop, thank you!

**Comment by MagicRB** [comment]
Repeated messages are frowned upon on GH, but I must thank you too. For both looking into this and keeping us in the loop. I and probably most, really appreciate it. Thanks again

**Comment by keslerm** [comment]
Excellent, thank you so much @tgross for the updates, it's extremely helpful to see the the progress on this.

**Comment by tgross** [comment]
I've built out a test rig that recreates the state I captured, and I've written a couple of tests that pass it through the `volumewatcher` and the GC job. And fortunately these tests fail with exactly the same behavior I witnessed yesterday.

I've opened https://github.com/hashicorp/nomad/pull/11890 with a draft approach. If a volume has any allocations that are nil, create a "past claim" on it when we _read_ the volume. This approach recognizes that we may never be able to fix all causes of desync between volumes and claims, because a node could potentially just up and disappear on us. This way trying to catch events from the clients is a matter of optimization and not correctness.

---

In manual testing of that patch, I discovered that we were not resetting the `volumewatcher`'s ACL token on leadership transitions. That means that even once we fix the claim state, the unpublish workflow will eventually start failing unrecoverably with errors like:

```
    2022-01-20T19:13:22.661Z [DEBUG] core.sched: forced volume claim GC
    2022-01-20T19:13:22.662Z [DEBUG] core.sched: CSI volume claim GC scanning before cutoff index: index=18446744073709551615 csi_volume_claim_gc_threshold=5m0s
    2022-01-20T19:13:22.664Z [ERROR] nomad.volumes_watcher: error releasing volume claims: namespace=default volume_id=csi-volume-nfs0
  error=
  | 1 error occurred:
  |     * Permission denied
  |
```

Or, if you detach manually the error is particularly confusing because you might think it's talking about _your_ ACL token:

```
    2022-01-20T19:12:57.559Z [ERROR] nomad.volumes_watcher: error releasing volume claims: namespace=default volume_id=csi-volume-nfs0
  error=
  | 1 error occurred:
  |     * Permission denied
  |
```

Either case leaves us in a state where volume claims have been released but the volumes are still physically mounted:

<details><summary>sad mount outputs</summary>

```
nomad-client2:~$ mount | grep nfs
192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 on /var/nomad/data/client/csi/node/org.democratic-csi.nfs/staging/csi-volume-nfs0/rw-file-system-single-node-writer type nfs4 (rw,noatime,vers=4.2,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.56.4,local_lock=none,addr=192.168.56.69)
192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 on /var/nomad/data/client/csi/node/org.democratic-csi.nfs/per-alloc/a1e552e3-e89d-9c8e-5e2a-0fe91a2041ec/csi-volume-nfs0/rw-file-system-single-node-writer type nfs4 (rw,noatime,vers=4.2,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.56.4,local_lock=none,addr=192.168.56.69)
192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 on /var/nomad/data/client/csi/node/org.democratic-csi.nfs/per-alloc/a90a1a1e-328c-e312-42b3-6766536306ea/csi-volume-nfs0/rw-file-system-single-node-writer type nfs4 (rw,noatime,vers=4.2,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.56.4,local_lock=none,addr=192.168.56.69)
```

</details>

I've opened https://github.com/hashicorp/nomad/pull/11891 with a fix for that.

---

Next steps for https://github.com/hashicorp/nomad/pull/11890:
* Do some more end-to-end testing to verify that it covers us in all the cases we can recreate.
* Make the `CSIVolumeDenormalize` method private and force all queries of volumes to hit it at the state store level, so that we ever can't miss doing so and end up playing wack-a-mole with this bug all over the code base.

---

Also, yesterday I mentioned:

> I've also got a patch going to do client-driven node unpublish and another moving claims from the client side to the plan applier, but that's probably not going to be ready till next week.

I made a brief attempt at moving claims into the plan applier, but it turned out to be _really_ awful, so that doesn't seem like the right approach after all. But the client-driven node unpublish looks feasible. I have a work-in-progress patch for that at https://github.com/hashicorp/nomad/pull/11892 but it doesn't quite work yet because the unpublish workflow doesn't tolerate it yet.

I'll be away tomorrow but will be picking all this up again on Monday.

**Comment by tgross** [comment]
I've spent much of today testing out the patch in https://github.com/hashicorp/nomad/pull/11890. In my testing I have yet to encounter this issue again once I've applied that patch (and https://github.com/hashicorp/nomad/pull/11891) to the cluster.

However, I'm still seeing volumes left attached unexpectedly as I described in https://github.com/hashicorp/nomad/issues/10927#issuecomment-1017920830. The `nomad volume detach` command doesn't seem to be working either, so my next step is to figure out if this patch breaks the unpublishing workflow unexpected or whether that's a separate problem.

---

I still haven't figured out what's going on with the access mode when releasing claims. It should be set to `<none>` so that the next claim can update it, but it's not. At this point I'm reasonably confident it's not tied to this problem though, so I'm going to split that out to a new issue: https://github.com/hashicorp/nomad/issues/11921

---

An interesting test case I encountered was purging a job and then immediately re-running the job. I was able to produce a case where the evaluation blocked because the allocations that were claiming writes had not yet unpublished the volume. But by the time the blocked eval was evaluated, the volume was free for scheduling and everything worked as expected without intervention. It looked like this:

```
$ nomad job stop -purge httpd
==> 2022-01-24T13:32:36-05:00: Monitoring evaluation "fbf2c487"
    2022-01-24T13:32:36-05:00: Evaluation triggered by job "httpd"
==> 2022-01-24T13:32:37-05:00: Monitoring evaluation "fbf2c487"
    2022-01-24T13:32:37-05:00: Evaluation within deployment: "cbe85da5"
    2022-01-24T13:32:37-05:00: Evaluation status changed: "pending" -> "complete"
==> 2022-01-24T13:32:37-05:00: Evaluation "fbf2c487" finished with status "complete"
==> 2022-01-24T13:32:37-05:00: Monitoring deployment "cbe85da5"
  ✓ Deployment "cbe85da5" successful

    2022-01-24T13:32:37-05:00
    ID          = cbe85da5
    Job ID      = httpd
    Job Version = 2
    Status      = successful
    Description = Deployment completed successfully

    Deployed
    Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
    web         2        2       2        0          2022-01-24T18:33:01Z

$ nomad job run ./jobs/csi/httpd.nomad
==> 2022-01-24T13:32:40-05:00: Monitoring evaluation "faae38d5"
    2022-01-24T13:32:40-05:00: Evaluation triggered by job "httpd"
==> 2022-01-24T13:32:41-05:00: Monitoring evaluation "faae38d5"
    2022-01-24T13:32:41-05:00: Evaluation within deployment: "5fcc4c00"
    2022-01-24T13:32:41-05:00: Evaluation status changed: "pending" -> "complete"
==> 2022-01-24T13:32:41-05:00: Evaluation "faae38d5" finished with status "complete" but failed to place all allocations:
    2022-01-24T13:32:41-05:00: Task Group "web" (failed to place 2 allocations):
      * Class "vagrant": 3 nodes excluded by filter
      * Constraint "CSI volume csi-volume-nfs0 has exhausted its available writer claims": 3 nodes excluded by filter
    2022-01-24T13:32:41-05:00: Evaluation "aa76c24a" waiting for additional capacity to place remainder
==> 2022-01-24T13:32:41-05:00: Monitoring deployment "5fcc4c00"
  ✓ Deployment "5fcc4c00" successful

    2022-01-24T13:33:04-05:00
    ID          = 5fcc4c00
    Job ID      = httpd
    Job Version = 0
    Status      = successful
    Description = Deployment completed successfully

    Deployed
    Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
    web         2        2       2        0          2022-01-24T18:42:52Z
```

This is actually working as intended, but the error message is confusing here. So I've added a commit onto https://github.com/hashicorp/nomad/pull/11890 that will specifically call out "hey this is a GC'd allocation and it should resolve itself momentarily".

**Comment by tgross** [comment]
Quick follow-up on this:

> However, I'm still seeing volumes left attached unexpectedly as I described in #10927 (comment). The `nomad volume detach` command doesn't seem to be working either, so my next step is to figure out if this patch breaks the unpublishing workflow unexpected or whether that's a separate problem.

There's some easy ways to end up with stray mounts if you node drain without `-ignore-system`, but (a) we're going to fix that separately, and (b) once I restored the node plugin I was _always_ able to `nomad node detach` in that case.

But I was able to eventually reproduce this by repeatedly purging a job and immediately re-running it with the access mode changed. The goal was to make it so that publishing was happening at roughly the same time we were trying to detach the volume. This ended up with a client that had the mount for a previous version of the job, but no alloc:

```
$ mount | grep nfs
192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 on /var/nomad/data/client/csi/node/org.democratic-csi.nfs/staging/csi-volume-nfs0/rw-file-system-single-node-writer type nfs4 (rw,noatime,vers=4.2,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.56.2,local_lock=none,addr=192.168.56.69)
192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 on /var/nomad/data/client/csi/node/org.democratic-csi.nfs/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer type nfs4 (rw,noatime,vers=4.2,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.56.2,local_lock=none,addr=192.168.56.69)
```

Here are the logs for the `NodePublishVolume` and `NodeUnpublishVolume` workflows on the node plugin:

<details><summary>NodePublishVolume</summary>

```json
{"level":"info","message":"new request - driver: ControllerNfsClientDriver method: NodeStageVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.42.0\"]},\"flags\":0},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"share\":\"/var/nfs/general/v/csi-volume-nfs0\",\"provisioner_driver\":\"nfs-client\",\"node_attach_driver\":\"nfs\",\"server\":\"192.168.56.69\"},\"volume_id\":\"csi-volume-nfs0\",\"staging_target_path\":\"/csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}}}","service":"democratic-csi"}
{"level":"debug","message":"operation lock keys: [\"volume_id_csi-volume-nfs0\"]","service":"democratic-csi"}
{"level":"verbose","message":"validating capabilities: [{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\",\"defaults\"],\"fs_type\":\"\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}]","service":"democratic-csi"}
executing filesystem command: stat /csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer
executing mount command: findmnt --source 192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 --mountpoint /csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount -t nfs -o noatime,defaults 192.168.56.69:/var/nfs/general/v/csi-volume-nfs0 /csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer
{"level":"info","message":"new response - driver: ControllerNfsClientDriver method: NodeStageVolume response: {}","service":"democratic-csi"}
{"level":"info","message":"new request - driver: ControllerNfsClientDriver method: NodePublishVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.42.0\"]},\"flags\":0},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"node_attach_driver\":\"nfs\",\"server\":\"192.168.56.69\",\"share\":\"/var/nfs/general/v/csi-volume-nfs0\",\"provisioner_driver\":\"nfs-client\"},\"volume_id\":\"csi-volume-nfs0\",\"staging_target_path\":\"/csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer\",\"target_path\":\"/csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"},\"readonly\":false}}","service":"democratic-csi"}
{"level":"debug","message":"operation lock keys: [\"volume_id_csi-volume-nfs0\"]","service":"democratic-csi"}
executing filesystem command: stat /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer
failed to execute filesystem command: stat /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer, response: {"code":1,"stdout":"","stderr":"stat: cannot stat '/csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer': No such file or directory\n","timeout":false}
executing filesystem command: mkdir -p -m 0750 /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer
executing mount command: findmnt --mountpoint /csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount --bind -o noatime,defaults /csi/staging/csi-volume-nfs0/rw-file-system-single-node-writer /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-single-node-writer
{"level":"info","message":"new response - driver: ControllerNfsClientDriver method: NodePublishVolume response: {}","service":"democratic-csi"}
```

</details>


<details><summary>NodeUnpublishVolume</summary>

```json
{"level":"info","message":"new request - driver: ControllerNfsClientDriver method: NodeUnpublishVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.42.0\"]},\"flags\":0},\"request\":{\"volume_id\":\"csi-volume-nfs0\",\"target_path\":\"/csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-multi-node-multi-writer\"}}","service":"democratic-csi"}
{"level":"debug","message":"operation lock keys: [\"volume_id_csi-volume-nfs0\"]","service":"democratic-csi"}
executing mount command: findmnt --mountpoint /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-multi-node-multi-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: stat /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-multi-node-multi-writer
failed to execute filesystem command: stat /csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-multi-node-multi-writer, response: {"code":1,"stdout":"","stderr":"stat: cannot stat '/csi/per-alloc/3f9bfe0b-0541-b83d-fd8b-a2b16ddcc318/csi-volume-nfs0/rw-file-system-multi-node-multi-writer': No such file or directory\n","timeout":false}
{"level":"info","message":"new response - driver: ControllerNfsClientDriver method: NodeUnpublishVolume response: {}","service":"democratic-csi"}
{"level":"info","message":"new request - driver: ControllerNfsClientDriver method: NodeUnstageVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.42.0\"]},\"flags\":0},\"request\":{\"volume_id\":\"csi-volume-nfs0\",\"staging_target_path\":\"/csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer\"}}","service":"democratic-csi"}
{"level":"debug","message":"operation lock keys: [\"volume_id_csi-volume-nfs0\"]","service":"democratic-csi"}
executing mount command: findmnt --mountpoint /csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer/block_device --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: stat /csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer
failed to execute filesystem command: stat /csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer, response: {"code":1,"stdout":"","stderr":"stat: cannot stat '/csi/staging/csi-volume-nfs0/rw-file-system-multi-node-multi-writer': No such file or directory\n","timeout":false}
{"level":"info","message":"new response - driver: ControllerNfsClientDriver method: NodeUnstageVolume response: {}","service":"democratic-csi"}
```

</details>

What we can see here is that the volume was mounted at `rw-file-system-single-node-writer` but when we're unpublishing we're doing so at `rw-file-system-multi-node-multi-writer`. So we'll never be able to unmount this because we're trying to unmount it at the wrong location. This is also why trying to detach it via `nomad node detach` results in a silent failure. We see the log for the HTTP request and then nothing else. We just aren't logging this code path.

So there's a race where a change in the access mode can land on the client during unpublish. There's a few obvious places that bug can be happening, so I'll pick that up again tomorrow.

**Comment by tgross** [comment]
Ok, the race described above turns out to be pretty straightforward. In the client's [`(*csiHook) Postrun()`](https://github.com/hashicorp/nomad/blob/v1.2.4/client/allocrunner/csi_hook.go#L97-L142) method, we make an unpublish RPC that includes a claim in the `CSIVolumeClaimStateUnpublishing` state and using the mode from the client. But then in the [`(*CSIVolume) Unpublish`](https://github.com/hashicorp/nomad/blob/v1.2.4/nomad/csi_endpoint.go#L562) RPC handler, we query the volume from the state store (because we only get an ID from the client). And when we make the client RPC for the node unpublish step, we use the _current volume's_ view of the mode (ref [`nodeUnpublishVolumeImpl`](https://github.com/hashicorp/nomad/blob/v1.2.4/nomad/csi_endpoint.go#L671-L672)). If the volume's mode has been changed before the old allocations can have their claims released, then we end up making a CSI RPC that will never succeed.

Why does this code path get the mode from the volume and not the claim? Because the claim written by the GC job in [`(*CoreScheduler) csiVolumeClaimGC`](https://github.com/hashicorp/nomad/blob/v1.2.4/nomad/core_sched.go#L723-L739) doesn't have a mode. Instead it just writes a claim in the unpublishing state to ensure the volumewatcher detects a "past claim" change and reaps all the claims on the volumes.

* With the patch in https://github.com/hashicorp/nomad/pull/11890 we should be able to safely ignore the case of the GC claim because all the past claims will be correct, so it should be safe to update `nodeUnpublishVolumeImpl` to always use the claims mode (maybe falling back to the volume's mode if unset?).
* We'll also fix most cases of this via https://github.com/hashicorp/nomad/pull/11892 because we'll drive the node unpublish step from the client which has the "real" view of what mode the volume was actually mounted with.
* I'll take a quick pass at auditing other uses of `vol.AccessMode` to ensure we're not making this mistake elsewhere.

**Comment by tgross** [comment]
* At this point I'm feeling pretty confident https://github.com/hashicorp/nomad/pull/11890 fixes the primary problem with the state store. I've just marked that as ready for review. 
* I've also pulled out https://github.com/hashicorp/nomad/pull/11931 which includes that patch but refactors the "past claims" logic so that we can be sure the core GC job and the volumewatcher are using the same code paths as other consumers of the state store. 
* That will unblock the last bit of code I need to finish up https://github.com/hashicorp/nomad/pull/11892 with the fix to the `Postrun()` race condition discussed this morning.

**Comment by tgross** [comment]
I've done some testing of https://github.com/hashicorp/nomad/pull/11892 and that's in good shape as well, and I've walked some team mates through the rest of the PRs so that we can get them reviewed and merged one-by-one. Unfortunately I've run into a non-CSI issue around the `stop_after_client_disconnect` flag https://github.com/hashicorp/nomad/issues/11943 which is going to impact a lot of CSI use cases (we created that flag specifically for stateful workloads). But we should be able to ship an incremental patch release with a bunch of the work done so far. Stay tuned for updates on when that may land as the PRs get merged.

**Comment by tgross** [comment]
The 4 patches for this issue (which also covers https://github.com/hashicorp/nomad/issues/10052 and https://github.com/hashicorp/nomad/issues/10833 and https://github.com/hashicorp/nomad/issues/8734) have been merged to `main`. We're working up a plan for a patch release shortly. Note this won't yet fix the issues in #8609 but it does make that an automatically recoverable situation. I'm going to be focusing heavily on knocking out the rest of the major CSI issues over the next month or two after we've shipped this.

Thanks again for your patience, folks.

**Comment by MagicRB** [comment]
We thank you dude! I can finally recommend nomad to people without a BUT, i've always wanted be able to do that.

**Comment by RickyGrassmuck** [comment]
This is awesome work @tgross! Highly appreciate the transparency and the depth at which you kept the community updated, this thread should become the gold standard for Organization <--> Community interactions!

**Comment by tgross** [comment]
Nomad 1.2.5 has shipped with the patches described here. We still have a few important CSI issues to close out before we can call CSI GA-ready, but it should be safe to close this issue now. Please let us know if you run into the issue with the current versions of Nomad. Thanks!

**Comment by m1keil** [comment]
@tgross I recently upgrade to v1.2.6 and I'm still hitting the issue.

After a few days of the job running, attempting to deploy ends up with the dreaded "CSI volume prometheus has exhausted...". Followed by a `nomad volume deregister --force` ritual.

**Comment by tgross** [comment]
@m1keil the problem isn't entirely eliminated until 1.3.0. https://github.com/hashicorp/nomad/pull/12113 should cover the remaining cases.

**Comment by iSchluff** [comment]
@tgross I have still managed to reproduce with v1.3.0-beta.1
After node restarts in my cluster a volume still got stuck. Maybe related: my cluster got stuck in a masterless state, but managed to recover itself after also updating the wan federated region to 1.3.0.

```
Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost  Unknown
registry    1       0         0        0       0         0     0

Placement Failure
Task Group "registry":
  * Constraint "CSI volume ceph-registry has exhausted its available writer claims and is claimed by a garbage collected allocation 45c884a0-e629-b180-eb44-c3c7a5d5e7b7; waiting for claim to be released": 3 nodes excluded by filter

Latest Deployment
ID          = 687ebec8
Status      = running
Description = Deployment is running
```
```
# nomad volume status ceph-registry
ID                   = ceph-registry
Name                 = ceph-registry
External ID          = 0001-0024-22b9524c-4a2d-4a56-9f3d-221ee05905a2-0000000000000001-af535f2b-9f9a-11ec-bf1c-0242ac110003
Plugin ID            = ceph-csi
Provider             = cephfs.csi.ceph.com
Version              = v3.6.0
Schedulable          = true
Controllers Healthy  = 1
Controllers Expected = 4
Nodes Healthy        = 3
Nodes Expected       = 3
Access Mode          = single-node-writer
Attachment Mode      = file-system
Mount Options        = <none>
Namespace            = default

Allocations
No allocations placed
```

**Comment by tgross** [comment]
> The node plugins couldn't reach the storage backend anymore and so the mount was still around but the task allocation was long gone. Whats the expected behavior here? 

A bit more detail would be helpful. Which task allocation do you mean? The plugin or the allocation that mounted the volume? I would not have expected the allocation that mounted the volume to be actually gone because we block for a [very long time](https://github.com/hashicorp/nomad/blob/v1.3.0-beta.1/client/allocrunner/csi_hook.go#L62) retrying the unmount _unless_ the client node was also shut down as well.

Did the node plugin for that node get restored to service? If so, the expectation is that the volumewatcher loop will eventually be able to detach it. Or is the client host itself gone too? 

> Maybe the volume allocation should be released after a long timeout has expired?

Unfortunately many storage providers will return errors if we do this. For example, imagine an AWS EBS volume attached to an EC2 instance where the Nomad client is shut down. We can never physically detach the volume until the client is restored, so all attempts to attach it elsewhere will fail unrecoverably. That's why we provide a `nomad volume detach` and `nomad volume delete` command, so that the operator can signal that the volume was detached out-of-band.

**Comment by iSchluff** [comment]
@tgross I mean the allocation that mounted the volume. It was gone because of a node restart and the rescheduled task remained in the stuck state since thursday, so approx. 4 days.

During the whole time the plugin state was shown as healthy in the ui, so nomad was not detecting that there was no actual connection to the backend. I guess ceph-csi just handles all Probe and GetCapabilities calls locally.

The network issue was eventually fixed, but the volume remained stuck.

> That's why we provide a `nomad volume detach` and `nomad volume delete` command, so that the operator can signal that the volume was detached out-of-band.
I haven't yet had much success with detach, because for me it normally just blocks and doesn't give out any information. For a case such as the one I described it is also hard to know which node the volume was attached to previously after the allocation has been gced. Would it even work if the node has been restarted?

But yes, if such a case can't be handled automatically it would be good if a volume could be unconditionally detached by the operator even if the node has been restarted or is gone altogether.
In my case I needed to deregister/reregister the volume again.

Also the deregister/reregister dance is impossible if the volume has been created with terraform, as the volume will be deleted automatically if the resource is removed.

**Comment by tgross** [comment]
> I mean the allocation that mounted the volume. It was gone because of a node restart and the rescheduled task remained in the stuck state since thursday, so approx. 4 days.
>
> During the whole time the plugin state was shown as healthy in the ui, so nomad was not detecting that there was no actual connection to the backend. I guess ceph-csi just handles all Probe and GetCapabilities calls locally.

Ok, so that would also have restarted the Node plugin tasks as well. When the node restarted, did the Node plugin tasks come back? The UI is maybe a little misleading here -- it shows the plugin as healthy if we can schedule volumes with it, not when 100% of plugins are healthy (otherwise we'd block volume operations during plugin deployments).

As for being stuck for 4 days... the evaluation for the rescheduled task hits a retry limit fairly quickly, so I suspect it's not "stuck" so much as "gave up." Once we've got the claim taken care of, re-running the job (or using [`nomad job eval`](https://www.nomadproject.io/docs/commands/job/eval)) should force a retry. But let's figure out the claim problem first. 😀 

> I haven't yet had much success with detach, because for me it normally just blocks and doesn't give out any information. 

It blocks because it's a synchronous operation; it has to make RPC calls to the Node plugin and to the Controller plugin. You should see logs on the Node plugin (first) and then the Controller plugin (second). If both the Node plugin and Controller plugin are live, that should work and you should be able to look at logs in both plugins (and likely the Nomad leader as well) to see what's going on there.

> For a case such as the one I described it is also hard to know which node the volume was attached to previously after the allocation has been gced. Would it even work if the node has been restarted?

So long as it restarted and has a running Node plugin, yes that should still work just fine. But fair point about knowing which node the volume was attached to previously.

**Comment by tgross** [comment]
(hit submit too soon  😀 )

> But yes, if such a case can't be handled automatically it would be good if a volume could be unconditionally detached by the operator even if the node has been restarted or is gone altogether.

It's not possible for us to unconditionally detach but in theory we could unconditionally free a claim. That's what [`deregister -force`](https://www.nomadproject.io/docs/commands/volume/deregister#force) is designed to do, but yeah I realize that's not going to work out so well with Terraform.
